frontend/views/containers/chatroom/ChatMain.vue

```diff
@@ -701,40 +702,59 @@ export default ({
       if (this.latestEvents.length >= CHATROOM_MAX_ARCHIVE_ACTION_PAGES * unit) {
-        sbp('gi.db/archive/delete', this.getArchiveKeyFromChatRoomId(chatRoomId))
+        sbp('gi.db/archive/delete', this.archiveKeyFromChatRoomId(chatRoomId))
       } else if (to !== this.messageState.prevTo || from !== this.messageState.prevFrom) {
         // this.currentChatRoomId could be wrong when the channels are switched very fast
         // so it's good to initiate using input parameter chatRoomId
-        sbp('gi.db/archive/save', this.getArchiveKeyFromChatRoomId(chatRoomId), JSON.stringify(this.latestEvents))
+        sbp('gi.db/archive/save', this.archiveKeyFromChatRoomId(chatRoomId), JSON.stringify(this.latestEvents))
       }
     },
-    getArchiveKeyFromChatRoomId (chatRoomId) {
+    archiveKeyFromChatRoomId (chatRoomId) {
       const curChatRoomId = chatRoomId || this.currentChatRoomId
-      return `messages/${curChatRoomId}`
+      return `messages/${this.ourUsername}/${curChatRoomId}`
     },
-    refreshContent: debounce(function (from, to) {
+    refreshContent: debounce(function (to, from) {
       // NOTE: using debounce we can skip unnecessary rendering contents
       this.archiveMessageState(from)
+      this.setMessageEventListener({ to, from })
       this.setInitMessages()
-      this.setMessageEventListener({ from, to })
     }, 250)
   },
   watch: {
-    'currentChatRoomId' (to, from) {
-      if (from) {
-        this.ephemeral.messagesInitiated = false
-      }
-      this.refreshContent(from, to)
-    },
-    'summary.isJoined' (to, from) {
-      if (to) {
+    'summary' (to, from) {
+      const toChatRoomId = to.chatRoomId
+      const fromChatRoomId = from.chatRoomId
+      const toIsJoined = to.isJoined
+      const fromIsJoined = from.isJoined
+
+      const initAfterSynced = (chatRoomId) => {
+        // TODO: `okTurtles.events/once` shouldn't be used here
+        //       since another contract could be started or finished being synced in advance
         sbp('okTurtles.events/once', CONTRACT_IS_SYNCING, (contractID, isSyncing) => {
-          if (contractID === this.currentChatRoomId && isSyncing === false) {
-            this.setMessageEventListener({ from: contractID, to: contractID })
+          if (contractID === chatRoomId && isSyncing === false) {
             this.setInitMessages()
           }
         })
       }
+
+      if (toChatRoomId !== fromChatRoomId) {
+        this.ephemeral.messagesInitiated = false
+        if (sbp('chelonia/contract/isSyncing', toChatRoomId)) {
+          this.archiveMessageState(fromChatRoomId)
+          this.setMessageEventListener({ to: toChatRoomId, from: fromChatRoomId })
+          if (toIsJoined && sbp('chelonia/contract/isSyncing', toChatRoomId)) {
+            initAfterSynced(toChatRoomId)
+          }
+        } else {
+          this.refreshContent(toChatRoomId, fromChatRoomId)
+        }
+      } else if (toIsJoined && toIsJoined !== fromIsJoined) {
```

Comment:
I don't understand this `isJoined` thing.

In `ChatMixins.js` it is set to `true`, so this condition doesn't make sense.